### PR TITLE
restart log stream call on INTERNAL_ERROR stream error

### PR DESCRIPTION
**Pull request by grantleehoffman**

### CODE REVIEW

**Review by jk-intuit — APPROVED** [review]

**Review by philnalwalker — APPROVED** [review]
lgtm

File: cli/internal/api/api.go [hunk 1]
```diff
@@ -74,8 +76,8 @@ func (c *Client) GetLogs(ctx context.Context, workflowName string) (responses.Ge
 	return output, nil
 }
 
-// StreamLogs streams the logs of a workflow.
-func (c *Client) StreamLogs(ctx context.Context, w io.Writer, workflowName string) error {
+// StreamLogs streams the logs of a workflow starting after loggedBytes.
+func (c *Client) StreamLogs(ctx context.Context, w io.Writer, workflowName string, loggedBytes int64) error {
```
Comment (szab100) [inline cli/internal/api/api.go:80]: shouldn't we use unsigned int? or this method should validate (>= 0)

File: cli/internal/api/api.go [hunk 2]
```diff
@@ -93,15 +95,22 @@ func (c *Client) StreamLogs(ctx context.Context, w io.Writer, workflowName strin
...
+	if _, err := io.CopyN(discardedWriter, resp.Body, loggedBytes); err != nil {
+		return err
+	}
+	loggedBytes, err = io.Copy(w, resp.Body)
 	if err != nil {
+		// retry call if we receive the stream error
+		if strings.Contains(err.Error(), "stream error: stream ID 1; INTERNAL_ERROR") {
+			time.Sleep(time.Second * 10)
+			// Restart log streaming
+			return c.StreamLogs(ctx, w, workflowName, loggedBytes)
```
Comment (szab100) [inline cli/internal/api/api.go:109]: shouldn't we use a counter to guarantee we give up after X retries?

File: cli/internal/api/api.go [hunk 3]
```diff
@@ -93,15 +95,22 @@ func (c *Client) StreamLogs(ctx context.Context, w io.Writer, workflowName strin
 		return fmt.Errorf("received unexpected status code: %d", resp.StatusCode)
 	}
 
-	_, err = io.Copy(w, resp.Body)
+	// discard reader bytes already logged
+	discardedWriter := &bytes.Buffer{}
```
Comment (szab100) [inline cli/internal/api/api.go:99]: is it better to use ioutil.Discard instead?

File: cli/internal/api/api.go [hunk 4]
```diff
@@ -93,15 +95,22 @@ func (c *Client) StreamLogs(ctx context.Context, w io.Writer, workflowName strin
...
+	if _, err := io.CopyN(discardedWriter, resp.Body, loggedBytes); err != nil {
+		return err
+	}
+	loggedBytes, err = io.Copy(w, resp.Body)
```
Comment (szab100) [inline cli/internal/api/api.go:103]: is it hard to cover this with tests? eg. mock Reader to return the INTERNAL_ERROR n-times with a test case that succeeds and another that exceeds max retries?

File: cli/cmd/logs.go [hunk 5]
```diff
@@ -22,11 +22,12 @@ var logsCmd = &cobra.Command{
 
 		apiCl := api.NewClient(argoCloudOpsServiceAddr(), "")
 
+		ctx := context.Background()
 		if streamLogs {
 			// This is a _very_ simple approach to streaming.
-			cobra.CheckErr(apiCl.StreamLogs(context.Background(), os.Stdout, workflowName))
+			cobra.CheckErr(apiCl.StreamLogs(ctx, os.Stdout, workflowName, 0))
```
[the remark this answers is not in the document]
Comment (szab100) [reply cli/cmd/logs.go:28]: (this is a only 'nit' comment, I know it's the "skip first N bytes" param.. it's just that in Java we usually create a constant like READ_FROM_START = 0, just to make it more readable)

File: cli/internal/api/api.go (hunk 1 above)
Comment (szab100) [reply cli/internal/api/api.go:80]: also maybe 'skipFirstBytes' or similar name would better express what this parameter is for?

**Pull request merged**